rdf-delta-server-local/src/main/java/org/seaborne/delta/zk/direct/ValidZooKeeperSupplier.java

```diff
@@ -0,0 +1,253 @@
+/*
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  See the NOTICE file distributed with this work for additional
+ *  information regarding copyright ownership.
+ */
+
+package org.seaborne.delta.zk.direct;
+
+import org.apache.zookeeper.*;
+import org.seaborne.delta.zk.ZkException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+/**
+ * Supplies valid, connected {@link ZooKeeper} instances or throws an exception.
+ *
+ * <p>
+ * {@link ZooKeeper} instances will attempt to reconnect on their own if the connection is lost. However, session
+ * expiration causes an instance to become unusable. This class monitors connection state changes and will instantiate
+ * a fresh {@link ZooKeeper} instance when a session expires. This class also monitors the Ensemble configuration and
+ * will update the {@link #connectString} to reflect dynamic changes to the Ensemble.
+ * </p>
+ * <p>
+ * If all the Ensemble members present in the configuration leave the ensemble while the {@link ZooKeeper} instance
+ * is invalid, new connection attempts will fail resulting in a {@link ZkException} that will crash the application.
+ * This crash is essential to enabling the detection and replacement of a failed RDF Delta node.
+ * </p>
+ * <p>
+ * If {@link #close()} is called, calls to {@link #get()} will yield a {@link ZkException} that should crash the
+ * system. Only call {@link #close()} when you are done with this {@link ValidZooKeeperSupplier}.
+ * </p>
+ */
+public final class ValidZooKeeperSupplier implements Supplier<ZooKeeper>, Watcher, AutoCloseable {
```

Comment:
I think `Supplier` is a crappy name, anyway. I was just reflecting that I implemented `Supplier` to signal to users that it can be used wherever a `Supplier<ZooKeeper>` is called for. What do you think of that rationale?

Reply:
A custom ZooKeeperProvider interface would be fine.